xlators/mgmt/glusterd/src/glusterd-op-sm.c

```diff
@@ -2830,6 +2832,151 @@ glusterd_set_shared_storage(dict_t *dict, char *key, char *value,
     return ret;
 }
 
+// Compare 2 strings for boolean value
+
+int
+glusterd_compare_bool_value(char *value_1, int value_2)
+{
+    int ret = -1;
+    gf_boolean_t bool_val_1 = _gf_false;
+
+    ret = gf_string2boolean(value_1, &bool_val_1);
+    if (ret == -1) {
+        gf_msg(THIS->name, GF_LOG_ERROR, 0, GD_MSG_INVALID_ENTRY,
+               "Invalid value entry");
+    }
+
+    if (!ret)
+        return bool_val_1 == value_2;
+    else
+        return ret;
+}
+
+// Compare values in the set in the volume/cluster with that of the predicate
+
+int
+glusterd_compare_option_value(glusterd_volinfo_t *volinfo, char *input_key,
+                              int input_value)
+{
+    struct volopt_map_entry *vme = NULL;
+    int ret = -1;
+    char *deflt_val = NULL;
+    int keylen;
+    glusterd_conf_t *priv = NULL;
+    xlator_t *this = THIS;
+
+    priv = this->private;
+    GF_VALIDATE_OR_GOTO(this->name, priv, out);
+
+    for (vme = &glusterd_volopt_map[0]; vme->key; vme++) {
+        if (!strcmp(vme->key, input_key)) {
+            // First look for the key in the priv->opts for global option
+            keylen = strlen(vme->key);
+            ret = dict_get_strn(priv->opts, vme->key, keylen, &deflt_val);
+            if (!deflt_val) {
+                // Check for the option in the volinfo dict
+                ret = dict_get_strn(volinfo->dict, vme->key, keylen,
+                                    &deflt_val);
+                if (ret == -ENOENT)
+                    // Check for glusterd default values
+                    deflt_val = glusterd_get_option_value(volinfo, vme->key);
+                if (!deflt_val) {
+                    if (vme->value) {
+                        deflt_val = vme->value;
+                    } else {
+                        // Check the option for tranlator default value
+                        ret = glusterd_get_value_for_vme_entry(vme, &deflt_val);
+                        if (ret)
+                            goto out;
+                        else if (ret == -2)
+                            continue;
+                    }
+                }
+            }
+        }
+    }
+
+    ret = glusterd_compare_bool_value(deflt_val, input_value);
+    deflt_val = NULL;
+
+out:
+    if (deflt_val)
+        GF_FREE(deflt_val);
+
```

Comment:
The problem with the above refactor is that the refactored function does not free _deflt_val_ and thus it's puts the burden of freeing on the caller.

To alleviate this, a variant should be added which is named `<function_name>_bool`, so tentatively, `glusterd_option_fetch_bool` It would wrap `glusterd_option_fetch` so that applies  `gf_string2boolean` to _deflt_val_, then frees _deflt_val_ and returns the resulting Boolean-or-Boolean-conversion-failure value (0, 1, or -1).